[트리] 11월 21일

jaeseo222 opened the pull request:
감사합니다!!!! 필수문제는,, 샘플코드 보며,,, 공부,,하겠습니다,,,!

Commit:
[트리] 11월 21일

감사합니다!!!! 필수문제는,, 샘플코드 보며,,, 공부,,하겠습니다,,,!

bsa0322 commented on 11월 16일 - 트리/14675.cpp:
```diff
@@ -0,0 +1,31 @@
+
+	int n, q;//n : 정점 개수, q : 질의 개수
+	int a, b;
+	vector<int> tree;
+
+	cin >> n;
+	tree.assign(n+1,0);
```
p3. 사소한 거긴 하지만 입력을 받고 벡터를 크기와 함께 선언해도 괜찮겠네요!!

bsa0322 commented on 11월 16일 - 트리/14675.cpp:
```diff
@@ -0,0 +1,31 @@
+			else cout << "no\n";
+
+		else//단절선이냐?
+			if (n > 1) cout << "yes\n";//총 노드 수가 2개 이상이기만 하면, 뭐든 단절선
+			else cout << "no\n";
```
p2. 문제에서 정점의 수인 n 의 범위를 다시 살펴보시면 해당 코드를 더 간결하게 만들 수 있어요! 
더불어 단절선과 단절점인지 판단하는 부분은 코드가 짧긴 하지만 함수화를 해서 사용해도 좋겠네요!!

bsa0322 commented on 11월 16일 - 트리/15681.cpp:
```diff
@@ -0,0 +1,54 @@
+	treeNum.assign(n + 1, 0);
+
+	//tree생성
+	n--;
+	while (n--) {
```
p3. 증감연산자의 위치를 적절히 배치하면 한 줄로 줄일 수 있어보여요!

bsa0322 commented on 11월 16일 - 트리/15681.cpp:
```diff
@@ -0,0 +1,54 @@
+	if (treeNum[root])//해당 정점에, subtree있을 시
+		return treeNum[root];//연결된 subtree의 정점 개수 리턴
+
+	//처음방문
+	visited[root] = true;
+	treeNum[root] = 1;//자신도 자신을 루트로 하는 서브트리에 포함되므로 1부터 시작
```
p3. treeNum 배열의 초기화값을 다르게 설정하면 visited를 처리하지 않고 풀 수 있는 방법이 있어요! 사실 트리는 부모-자식 관계이기에 항상 부모노드만 걸러주면 되는데, 이때 부모노드를 방문한다 해도 treeNum의 값이 변하지 않으려면 어떤 수로 초기화를 해야 할까요!? 
하지만 사실 지금도 잘 풀어주셔서 제가 지금 말씀드린 부분은 나중에 샘플코드 참고해보셔도 좋을 것 같아요 😊

bsa0322 commented on 11월 16일 - 트리/1967.cpp:
```diff
@@ -0,0 +1,52 @@
+#include<iostream>
+#include<vector>
+using namespace std;
+
+const int SIZE = 10000 + 1;
```
p3. 10의 제곱수는 1e(지수) 형태로 표현하는 것도 편하고 좋아요!!

bsa0322 commented on 11월 16일 - 트리/1967.cpp:
```diff
@@ -0,0 +1,52 @@
+const int SIZE = 10000 + 1;
+vector<pair<int, int>> tree[SIZE];//first: 자식, second: 간선 가중치(길이)
+vector<bool> visited;
+int dia = 0, far = 0;//지름, 가장 멀리있는 점
+
+void dfs(int n, int len) {
+	if (visited[n])//방문했으면 리턴
+		return;
+
+	//처음방문
+	visited[n] = true;
+
+	if (dia < len) far = n;	//가장 멀리있는 점 갱신
+	dia = max(dia, len);	//지름 갱신
+
+	for (int i = 0; i < tree[n].size(); i++) {
+		int child = tree[n][i].first;//자식
+		int weight = len + tree[n][i].second;//길이 = 현재 길이 + 자식 가중치
+
+		dfs(child, weight);
+	}
+}
```
p2. 리턴값이 있는 dfs를 사용하면 전역변수의 사용을 줄일 수 있어요! 메인에서 필요한 전역변수 2개 모두 dfs를 통해서 값을 얻을 수 있네요! 리턴값이 있다는 건 어떻게 보면 함수를 호출한 후, return;을 해준 것과 비슷한데, 그냥 return; 하는게 아니라 함수에서 계산한 값을 가져간다 생각해도 좋아요..!

bsa0322 commented on 11월 16일 - 트리/5639.cpp:
```diff
@@ -0,0 +1,31 @@
+	int input, idx = 0;
+	int tree[SIZE];
+
+	//입력
+	while (true) {
+		cin >> input;
+		if (cin.eof()) break;
+		tree[idx++] = input;
+	}
```
p3. while 조건문을 활용해서 입력을 받는 방법도 있어요. 만약 더 이상 입력이 없다면 cin.eof()가 true를 반환하는데, 반대로 입력이 있을 동안에 cin >> input은 어떤 값을 가질까요?

jaeseo222 replied:
엄청 신기해요!

bsa0322 reviewed (approved):
p2. 수고하셨습니다!! 잘 풀어주셔서 제가 크게 코멘트 드릴 부분은 없었네요!! 남긴 코멘트들은 수정해보셔도 좋고, 머지해주셔도 괜찮아요! 수정하신다면, 저 다시 리뷰어로 호출해주세요. 수고하셨습니다~! 👍

Commit:
[트리] 11월 22일 - Update

친절하고, 효율적인 코드 설명 감사드려요!!! 그런데, 1967번 설명을 보충해주실 수 있으실까용?! 어떻게 전역변수를 사용하지 않을 수 있을지 와닿지 않아서요!!

bsa0322 commented on 11월 16일 - 트리/15681.cpp:
```diff
@@ -0,0 +1,47 @@
+vector<int> treeNum;
+
+//dfs
+int dfs(int root) {
+	if (treeNum[root])//해당 정점에, subtree있을 시
+		return 0;//더할 필요 없으니 0 리턴
+
+	//처음방문
+	treeNum[root] = 1;//자신도 자신을 루트로 하는 서브트리에 포함되므로 1부터 시작
+	for (int i = 0; i < tree[root].size(); i++)
+		treeNum[root] += dfs(tree[root][i]);//자식노드로 들어감
+
+	return treeNum[root];//현재까지 subtree 개수 리턴
+}
```
헐 제가 처음 제안드린 방법과 다르긴 한데..! 잘 풀어주셨네요!! 
그런데 사실 지금 subtree가 있을 때 0을 리턴하는 게 어떻게 되는건지 너무 신기해요.....

jaeseo222 replied:
헤헷

bsa0322 reviewed (commented):
수정해주신 부분 확인했습니다! 좋아요~!! 머지해주셔도 됩니다! 👍
